### DISCUSSION TITLE
Fixing Akkling (F#) cluster sharding support broken after Akka.NET 1.5.15 release

### DISCUSSION THREAD

**Discussion by object** [akkadotnet/akka.net#7058] [Q&A] [answered]
Akka.NET 1.5.15 comes with an updated IMessageExtractor interface optimized to give better performance for computing ShardId. Akkling open source library has an internal type TypedMessageExtractor that needs to be updated for the revised IMessageExtractor interface. Here' the previous version of TypedMessageExtractor:

```
type internal TypedMessageExtractor<'Envelope, 'Message>(extractor: 'Envelope -> string*string*'Message) =
    interface IMessageExtractor with
        member this.ShardId message =
            match message with
            | :? 'Envelope as env -> 
                let shardId, _, _ = extractor(env)
                shardId
            | _ -> null
        member this.EntityId message =
            match message with
            | :? 'Envelope as env -> 
                let _, entityId, _ = extractor(env)
                entityId
            | _ -> null
        member this.EntityMessage message =
            match message with
            | :? 'Envelope as env -> 
                let _, _, msg = extractor(env)
                box msg
            | _ -> null

```

TypedMessageExtractor constructor takes a parameter **extractor** which is a function that for a given instance of **Envelope** returns a tuple of EntityId, ShardId and Message. This no longer fits the new IMessageExtractor interface which has an overload computing ShardId from EntityId. But the overload has an extra parameter **messageHint** and I supposed it can be used as a fallback. Here's a proposed workaround to make TypedMessageExtractor compatible with the new IMessageExtractor. New method:
```
        member this.ShardId (entityId, messageHint) =
            match messageHint with
            | :? 'Envelope as env -> 
                let shardId, _, _ = extractor(env)
                shardId
            | _ -> null

```

As you see, the **entityId** parameter is not used in the implementation, instead extractor function is applied to a **messageHint** parameter. I tested it in our project, and it seems to work. Is this a sufficient workaround to have a maintenance release of Akkling library that will work with Akka.NET 1.5.15?

**Comment by Aaronontheweb** [comment, marked as answer]
The `messageHint` variable was designed for exactly this type of scenario - to add backwards compat for cases where the `ShardId` needs to be computed from the message (i.e. working with the `ExtractShardId` delegate in Akka.NET, which has this limitation.)

So yes @object I think that work-around should work in lieu of a maintenance update for Akkling. Does that help answer your question?

**Reply by object** [reply]
Thanks for the super-quick response! Yes, this is the answer I wanted. In the meantime I will try to reach Bartosz and think about a better approach to take full advantage of the new IMessageExtractor.
